index.bs

```diff
@@ -926,9 +951,25 @@ Geometric Primitives {#geometricprimitives}
 Matrices {#matrices}
 --------
 
-WebXR provides various transforms in the form of <dfn lt="matrix">matrices</dfn>. WebXR matrices are always 4x4 and given as 16 element {{Float32Array}}s in column major order. They may be passed directly to WebGL's {{uniformMatrix4fv}} function, used to create an equivalent {{DOMMatrix}}, or used with a variety of third party math libraries.
+WebXR provides various transforms in the form of <dfn lt="matrix">matrices</dfn>. WebXR uses the WebGL conventions when communicating matrices, in which 4x4 matrices are given as 16 element {{Float32Array}}s with column major storage, and are applied to column vectors by premultiplying the matrix from the left. They may be passed directly to WebGL's {{uniformMatrix4fv}} function, used to create an equivalent {{DOMMatrix}}, or used with a variety of third party math libraries.
```

Comment:
As far as I can tell, the core WebGL is silent on this since it doesn't discuss matrix math at all. There are a few references to column-major vs row-major layout qualifiers and a UNIFORM_IS_ROW_MAJOR attribute, but the actual usage is in shaders, and it's largely up to the user how this is intended to be interpreted.

GLSL does specify column-major matrix storage order in various places, i.e. 4.1.6 "Matrices are read from and written to in column major order"

 5.11 "Vector and Matrix Operations" describes how both "v * m" and "m * v" "perform the correct linear algebraic multiply", but the descriptions are quite low-level and don't use the terms "row vector", "column vector", or pre/postmultiply. I think it's compatible with our usage, but I find it a bit harder to unpack what's going on.

https://www.khronos.org/registry/webgl/specs/latest/2.0/
https://www.khronos.org/registry/OpenGL/specs/es/2.0/GLSL_ES_Specification_1.00.pdf

Reply:
I think there's nothing stopping an app developer from writing a WebGL app based on their own choice of matrix conventions, i.e. using row vectors with postmultiplied matrices, since all the needed operations are available and it'll work as long as they are consistent about it. So I think it's fair that the WebGL and GLSL specs don't specifically require a certain convention.

By contrast, WebXR is more opinionated about this due to actually supplying matrices, so we need to be clear about the convention. In the (unlikely?) case that this doesn't match how an app developer uses matrices, they'll at least have the necessary information how to do conversions as needed.

Reply:
The WebGL spec doesn't actually define them. It relies on the OpenGL ES 2.0 spec to do that, which is a giant honkin PDF with zero linkability. 😠

Reply:
Specifically, does this matrix representation match the matrix representation we see used in three.js, A-Frame, Babylon, etc.?

One of the reasons we avoided giving matrices back directly from OpenXR was that engines widely diverged on their matrix conventions, and so most apps would need to convert anyway - position/orientation was much less ambiguous.  In practice, has the web community centralized on a single matrix standard here that we are now embracing?

Reply:
overall graphics _notation_ seems to have settled on column vectors (i.e. premultiply vectors by transforms).

matrix representations seem to differ between row major and column major in software (column major apparently has SIMD benefits, row major matches how it looks typed out)

I feel like for matrices in JS everyone uses glMatrix which is column-major? Maybe I'm wrong.

I'm becoming more convinced that we should just avoid exposing `matrix` and talk in terms of pos/orientation only (all the algorithms can be rewritten to work here), despite strongly disagreeing with that earlier. We also have the transform from vs transform to distinction that muddies up what `matrix` precisely is.